tests/wpunit/Admin/Fields/FieldTest.php

```diff
@@ -0,0 +1,216 @@
+		$this->assertEquals( 'License key', $field->get_placeholder() );
+		$this->assertEquals( 'stellarwp-uplink-license-key-field', $field->get_classes() );
+	}
+
+	/**
+	 * @test
+	 * @dataProvider resourceProvider
+	 */
+	public function it_should_render_fields_correctly( $resource ) {
+		$current_resource = $this->setup_container_get_slug( $resource );
+		$slug             = $current_resource->get_slug();
+		$field            = $this->container->get( Field::class )->set_resource( $current_resource );
+		$license_key      = 'license_key' . $slug;
+		$option_name      = $current_resource->get_license_object()->get_key_option_name();
+		$this->set_fn_return( 'wp_create_nonce', '123456789', false );
+
+		// Update the license key to a known value.
+		update_option( $option_name, $license_key );
+		$option_value = get_option( $option_name );
+		$this->assertEquals( $option_value, $license_key );
+
+		$field_name = 'field-' . $slug;
+		$field->set_field_name( $field_name );
+
+		$test = $field->get_render_html();
+		$this->assertMatchesHtmlSnapshot( $test );
+	}
+
+
+	/**
+	 * @test
+	 * @dataProvider resourceProvider
+	 */
+	public function it_should_set_and_get_field_id( $resource ) {
+		$field = $this->container->get( Field::class )->set_resource( $this->setup_container_get_slug( $resource ) );
+
+		$field_id = 'custom-field-id';
+		$field->set_field_id( $field_id );
+
+		$this->assertEquals( $field_id, $field->get_field_id() );
+	}
+
+	/**
+	 * @test
+	 * @dataProvider resourceProvider
+	 */
+	public function it_should_set_and_get_field_label( $resource ) {
+		$field = $this->container->get( Field::class )->set_resource( $this->setup_container_get_slug( $resource ) );
+
+		$label = 'Custom Label';
+		$field->set_label( $label );
+
+		$this->assertEquals( $label, $field->get_label() );
+	}
+
+	/**
+	 * @test
+	 * @dataProvider resourceProvider
+	 */
+	public function it_should_get_placeholder( $resource ) {
+		$field = $this->container->get( Field::class )->set_resource( $this->setup_container_get_slug( $resource ) );
+
+		$this->assertEquals( 'License key', $field->get_placeholder() );
+	}
+
+	/**
+	 * @test
+	 * @dataProvider resourceProvider
+	 */
+	public function it_should_get_nonce_action_and_field( $resource ) {
+		$field = $this->container->get( Field::class )->set_resource( $this->setup_container_get_slug( $resource ) );
+
+		$nonce_action = $this->container->get( Group::class )->get_name();
+		$nonce_field  = $field->get_nonce_field();
+
+		// Extract the nonce value from the nonce field
+		preg_match( '/value=["\']([^"\']+)["\']/', $nonce_field, $matches );
+		$nonce_value = $matches[1];
+
+		$this->assertNotEmpty( $nonce_action, 'Nonce action should not be empty.' );
+		$this->assertStringContainsString( 'stellarwp-uplink-license-key-nonce__' . $field->get_slug(), $nonce_field, 'Nonce field should contain the correct action slug.' );
+
+		// Validate the nonce
+		$is_valid_nonce = wp_verify_nonce( $nonce_value, $nonce_action );
```

Comment:
Both this test case and the `UplinkTestCase` do not set the current user to one that can, say, manage options (i.e. an administrator); is there any (or should there be any) `current_user_can` check in the field logic to avoid output when the user should not be entitled to see the field?

Reply:
@lucatume great question. I think we haven't discussed it yet and i am going to bring it up. My logic says that user capabilities are to be managed externally of the uplink library. If they are to be managed within, it should be configurable so it can serve all the different needs.